### PR TITLE
composition: implement composite key support

### PR DESCRIPTION
**Pull request by tomhoule**
# Description

Concretely, entities defining keys like `@key(fields: "id name")` and
even nested selections like `@key(fields: "patient { id } recordId")`.

A lot of the complexity in this PR comes from the machinery to detect
that:

1. An object is part of an entity in a subgraph because one of its field
   is part of a nested key on another entity.
2. A field is part of a key, not on its parent object but another object
   that includes it in its key through a nested selection.

closes GB-5264

# Type of change

- [ ] 💔 Breaking
- [x] 🚀 Feature
- [ ] 🐛 Fix
- [ ] 🛠️ Tooling
- [ ] 🔨 Refactoring
- [ ] 🧪 Test
- [ ] 📦 Dependency
- [ ] 📖 Requires documentation update

### CODE REVIEW

File: engine/crates/composition/src/compose/object.rs [hunk 1]
```diff
@@ -45,16 +45,16 @@ pub(super) fn compose_object_fields<'a>(first: FieldWalker<'a>, fields: &[FieldW
         ));
     }
 
-    let first_is_key = first.is_key();
-    if fields.iter().any(|field| field.is_key() != first_is_key) {
+    let first_is_key = first.is_part_of_key();
```
Comment (tomhoule) [inline engine/crates/composition/src/compose/object.rs:48]: should be renamed to `first_is_part_of_key`

File: engine/crates/composition/src/subgraphs/fields.rs [hunk 2]
```diff
@@ -100,23 +100,14 @@ impl<'a> FieldWalker<'a> {
         self.field().arguments.iter().map(move |id| self.walk(*id))
     }
 
-    /// Returns true iff there is an `@key` directive containing exactly this field (no composite
-    /// key).
-    pub fn is_key(self) -> bool {
+    /// Returns true iff there is an `@key` directive containing this field, possibly with others
+    /// as part of a composite key.
+    pub(crate) fn is_part_of_key(self) -> bool {
```
Comment (tomhoule) [inline engine/crates/composition/src/subgraphs/fields.rs:105]: Note that there's nothing about nesting in the implementation. What does it mean for a nested field to be part of the key (in another value type or entity)? For example:

```graphql
# First subgraph

type A @key("b { id } c") {
  b: B!
  c: String!
}

type B {
  id: ID!
  foo: String
}

# Other subgraph

type B @key("id") {
  id: ID!
  bar: String
}
```

Is that valid?

Comment (tomhoule) [reply engine/crates/composition/src/subgraphs/fields.rs:105]: Apparently it should be.

Comment (tomhoule) [reply engine/crates/composition/src/subgraphs/fields.rs]: Implemented.

**Comment by tomhoule** [comment]
The PR could use more tests, there's a lot of edge cases when you start to consider nested keys.

File: engine/crates/composition/src/ingest_subgraph/nested_key_fields.rs [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+                    };
+
+                    for subselection_field in &field.subselection {
+                        ingest_nested_key_fields_rec(field_type, subselection_field, nested_key_fields);
+                    }
+                }
+            }
+        }
+    });
+}
```
Comment (pimeys) [inline engine/crates/composition/src/ingest_subgraph/nested_key_fields.rs:23]: What about:

```suggestion
pub(super) fn ingest_nested_key_fields(subgraph_id: SubgraphId, subgraphs: &mut Subgraphs) {
    subgraphs.with_nested_key_fields(|subgraphs, nested_key_fields| {
        let keys = subgraphs
            .walk(subgraph_id)
            .definitions()
            .flat_map(|definition| definition.entity_keys());

        for key in keys {
            for field in key.fields() {
                let Some(field_type) = key
                    .parent_definition()
                    .find_field(field.field)
                    .and_then(|field| field.r#type().definition(subgraph_id))
                else {
                    continue;
                };

                for subselection_field in &field.subselection {
                    ingest_nested_key_fields_rec(field_type, subselection_field, nested_key_fields);
                }
            }
        }
    });
}
```

You could make this even better if `KeyWalker::fields` would return an iterator of walkers, and `Selection` could know in which key it's in.

File: engine/crates/composition/src/subgraphs/definitions.rs [hunk 4]
```diff
@@ -90,3 +90,17 @@ impl<'a> DefinitionWalker<'a> {
         self.walk(self.definition().subgraph_id)
     }
 }
+
+impl<'a> SubgraphWalker<'a> {
```
Comment (pimeys) [inline engine/crates/composition/src/subgraphs/definitions.rs:94]: Btw, why we want to split these implementations to different modules? Now it's a bit tricky to have a full vision what is implemented in `SubgraphWalker`, if it's in multiple places.

File: engine/crates/composition/src/subgraphs/definitions.rs [hunk 5]
```diff
@@ -90,3 +90,17 @@ impl<'a> DefinitionWalker<'a> {
...
+    pub(crate) fn definitions(self) -> impl Iterator<Item = DefinitionWalker<'a>> {
+        let subgraph_id = self.id;
+        let definitions = &self.subgraphs.definitions.0;
+        let start = definitions.partition_point(|def| def.subgraph_id < self.id);
+        let subgraph_definitions = definitions[start..]
```
Comment (pimeys) [inline engine/crates/composition/src/subgraphs/definitions.rs:99]: This stuff smells like something we could have in a more generic way available somewhere.

File: engine/crates/composition/src/subgraphs/keys.rs [hunk 6]
```diff
@@ -61,6 +65,44 @@ impl Subgraphs {
...
+    /// type Nested {
+    ///   identifier: ID!
+    /// }
+    ///
+    /// ```
+    ///
+    /// `Nested.identifier` is a nested key field.
+    fields: HashSet<FieldId>,
+
+    /// Objects that are part of keys that are not defined on the object itself.
```
Comment (pimeys) [inline engine/crates/composition/src/subgraphs/keys.rs:97]: Can an object be defined on the object itself? Just asking, this reads a bit weird.

File: engine/crates/composition/src/subgraphs/keys.rs [hunk 7]
```diff
@@ -97,12 +139,39 @@ impl<'a> KeyWalker<'a> {
...
 
 impl<'a> DefinitionWalker<'a> {
+    pub fn is_entity(self) -> bool {
+        self.entity_keys().next().is_some()
+            || self
+                .subgraphs
+                .keys
+                .nested_key_fields
+                .objects_with_nested_keys
+                .contains(&self.id)
```
Comment (pimeys) [inline engine/crates/composition/src/subgraphs/keys.rs:149]: This stuff is ripe for some apis on a bit higher level, so you don't need to traverse the tree so deep to ask that question.

File: engine/crates/composition/src/subgraphs/keys.rs [hunk 8]
```diff
@@ -97,12 +139,39 @@ impl<'a> KeyWalker<'a> {
...
+
     pub fn entity_keys(self) -> impl Iterator<Item = KeyWalker<'a>> {
-        let start = self.subgraphs.keys.0.partition_point(|(parent, _)| *parent < self.id);
-        self.subgraphs.keys.0[start..]
+        let start = self
+            .subgraphs
+            .keys
+            .keys
+            .partition_point(|(parent, _)| *parent < self.id);
+        self.subgraphs.keys.keys[start..]
```
Comment (pimeys) [inline engine/crates/composition/src/subgraphs/keys.rs:158]: Again, there's a possibility for some generic function that could handle the partitioning.

**Review by pimeys — APPROVED** [review]
Nitpicks, but good stuff.

File: engine/crates/composition/src/subgraphs/definitions.rs (hunk 4 above)
Comment (tomhoule) [reply engine/crates/composition/src/subgraphs/definitions.rs:94]: I'm experimenting with this because it lets me keep the data in the `subgraphs::keys` module private (for example), so everything interacting with that piece of data is encapsulated in the module. Jury is still out on whether that's a better approach, but the crate is structured like that at the moment.

File: engine/crates/composition/src/subgraphs/definitions.rs (hunk 5 above)
Comment (tomhoule) [reply engine/crates/composition/src/subgraphs/definitions.rs:99]: See `RangeWalker` in @Finistere 's introspection PR.

File: engine/crates/composition/src/subgraphs/keys.rs (hunk 6 above)
Comment (tomhoule) [reply engine/crates/composition/src/subgraphs/keys.rs:97]: I can add a code example.

File: engine/crates/composition/src/subgraphs/keys.rs (hunk 7 above)
Comment (tomhoule) [reply engine/crates/composition/src/subgraphs/keys.rs:149]: It's nested because we're defining methods on `Subgraphs`, but we're in the module that defines NestedKeyFields.

File: engine/crates/composition/src/ingest_subgraph/nested_key_fields.rs (hunk 3 above)
Comment (tomhoule) [reply engine/crates/composition/src/ingest_subgraph/nested_key_fields.rs:23]: `definition` would be out of scope, unless you add a map inside the flat map. Selection is not doable because it's not normalized so you can't have unique identifiers, it was discussed in a previous review with Benjamin.
